TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/ExpiringTrafficStreamMap.java

```diff
@@ -0,0 +1,356 @@
+            behavioralPolicy.onNewDataArrivingAfterItsAccumulationHasBeenExpired(partitionId, connectionId,
+                    lastPacketTimestamp.toInstant(), expiringQueue.getLatestPossibleKeyValue());
+            return false;
+        }
+        if (timestampMillis.test(lastPacketTimestamp, (newTs, lastTs) -> newTs > lastTs)) {
+            var witnessValue = accumulation.newestPacketTimestampInMillis.compareAndExchange(lastPacketTimestamp.millis,
+                    timestampMillis.millis);
+            if (lastPacketTimestamp.millis != witnessValue) {
+                ++attempts;
+                if (behavioralPolicy.shouldRetryAfterAccumulationTimestampRaceDetected(partitionId, connectionId,
+                        timestampMillis.toInstant(), accumulation, attempts)) {
+                    return updateExpirationTrackers(partitionId, connectionId, timestampMillis,
+                            accumulation, attempts);
+                } else {
+                    return false;
+                }
+            }
+        }
+
+        var targetBucketHashSet = expiringQueue.getHashSetForTimestamp(timestampMillis);
+        if (targetBucketHashSet == null) {
+            var startOfWindow = expiringQueue.firstKey().toInstant();
+            assert !timestampMillis.test(startOfWindow, (ts, windowStart) -> ts < windowStart) :
+                    "Only expected to not find the target bucket when the incoming timestamp was before the " +
+                            "expiring queue's time window";
+            behavioralPolicy.onDataArrivingBeforeTheStartOfTheCurrentProcessingWindow(partitionId, connectionId,
+                    timestampMillis.toInstant(), expiringQueue.getLatestPossibleKeyValue());
+            return false;
+        }
+        if (lastPacketTimestamp.millis > ACCUMULATION_TIMESTAMP_NOT_SET_YET_SENTINEL) {
+            var sourceBucket = expiringQueue.getHashSetForTimestamp(lastPacketTimestamp);
+            if (sourceBucket != targetBucketHashSet) {
+                if (sourceBucket == null) {
+                    behavioralPolicy.onNewDataArrivingAfterItsAccumulationHasBeenExpired(partitionId, connectionId,
+                            timestampMillis.toInstant(), expiringQueue.getLatestPossibleKeyValue());
+                    return false;
+                }
+                // this will do nothing if it was already removed, such as in the previous recursive run
+                sourceBucket.remove(connectionId);
+            }
+        }
+        targetBucketHashSet.put(connectionId, Boolean.TRUE);
+        return true;
+    }
+
+    Accumulation get(String partitionId, String connectionId, Instant timestamp) {
+        var accumulation = connectionAccumulationMap.get(new ScopedConnectionIdKey(partitionId, connectionId));
+        if (accumulation == null) {
+            return null;
+        }
+        if (!updateExpirationTrackers(partitionId, connectionId, new EpochMillis(timestamp), accumulation, 0)) {
```

Comment:
Correct. I'm open to suggestions on how to do an out of band flush. I'm thinking that it will be even messier and maybe better in a subsequent PR (I should make a comment and Jira task)